### PR TITLE
[Frontend,MLIR] Improve Compiler driver error reporting

### PR DESCRIPTION
**Pull request by grwlf**
The goal of this PR is

* [x] Python should report a compiler error, rather than simply RuntimeError: Compilation failed
* [x] The error message should contain the compilation stage (pass) that failed.
* [x] The error message should contain MLIR debug info if available, currently only the MLIR error without context is shown: expected mixed offsets rank to match mixed sizes rank (0 vs 2) so the rank of the result type is well-formed.
  + Note: the mentioned error is raised from [Here](https://github.com/llvm/llvm-project/blob/4706251a3186c34da0ee8fd894f7e6b095da8fdc/mlir/lib/Interfaces/ViewLikeInterface.cpp#L48)
  + Verbose mode now includes MLIR stack traces

In the course of work, the following issues were also found and fixed:

* [x] Fixed CompilerDriver failures on empty pipelines
* [x] Compiler class no longer ignores custom pipelines in textual_ir mode
* [x] Fixed the name-ordering among the pipeline dump files

This PR adds a small test pass which might be a controversial decision due to a large number of CMakeLists modifications required. Currently the pass always fails with an error message - we use it to test the error reporting. MLIR upstream seems to also practice creating test passes for dialects, so I think we can re-use it later as well. 

[sc-46580]

### CODE REVIEW

File: frontend/catalyst/compiler.py [hunk 1]
```diff
@@ -355,15 +361,18 @@ def run_from_ir(
...
+                ir,
+                workspace,
+                module_name,
+                keep_intermediate=self.options.keep_intermediate,
+                verbose=self.options.verbose,
+                pipelines=pipelines,
+                lower_to_llvm=lower_to_llvm,
+            )
+        except RuntimeError as e:
+            raise CompileError(*e.args) from e
```
Comment (erick-xanadu) [inline frontend/catalyst/compiler.py:375]: Is there a particular advantage to change this from a `RuntimeError` to a `CompileError` error? Just curious,

File: mlir/lib/Driver/CompilerDriver.cpp [hunk 2]
```diff
@@ -274,79 +286,108 @@ LogicalResult runEnzymePasses(const CompilerOptions &options,
...
+        for (auto pipelineName : danglingPipelines) {
+            {
+                llvm::raw_string_ostream s{outputs[pipelineName]};
+                s << *moduleOp;
+            }
+            dumpToFile(options, output.nextPipelineDumpFilename(pipelineName),
+                       outputs[pipelineName]);
+        }
+    }
+
```
Comment (erick-xanadu) [inline mlir/lib/Driver/CompilerDriver.cpp:390]: This printing process seems a bit more complicated than needed. Is it possible to do it differently?

What is a dangling pipeline?

**Review by erick-xanadu — COMMENTED** [review]
Cool! I am not a fan of creating a pass that just fails in order to test the reporting framework. I think it is enough to have something similar to the "compile from IR" error where that it is guaranteed to fail at compile time. As to what is something that is guaranteed to fail at compile time that cannot be detected by a simple parse? I am not too sure. So perhaps, this test is indeed needed.

I also feel like the printing mechanism has gotten a bit convoluted. What are dangling pipelines? It also looks like the print handler is called after each single pass inside the pipeline as opposed to just call the print handler at the end of the pipeline:

```c++
    // For each pipeline-terminating pass, print the IR into the corresponding dump file and
    // into a diagnostic output buffer. Note that one pass can terminate multiple pipelines.
    auto afterPassCallback = [&](Pass *pass, Operation *op) {
        if (!options.keepIntermediate)
            return;
        auto res = pipelineTailMarkers.find(pass);
        if (res != pipelineTailMarkers.end()) { // <-- only if it is the end it is printed
        // but the fact that we are in the printer callback is a bit strange to me.
```

I am thinking something that instead looks like:

```c++
for pipeline {
  // prepare pass manager
   if (failed(pm.run(moduleOp))) {
        return failure();
    }
    print moduleOp
}
```

This is already being done for the LLVM pipelines. What do you think? Is it possible? I also feel like this would enable us to use multithreading since it appears that multithreading is disabled when printing using the IR printer as opposed to calling print on the module itself.

File: frontend/catalyst/compiler.py (hunk 1 above)
Comment (grwlf) [reply frontend/catalyst/compiler.py:375]: I read the first point of the task description as a direct instruction to raise CompilerError. But I think I might be wrong, will clarify this

File: mlir/lib/Driver/CompilerDriver.cpp (hunk 2 above)
Comment (grwlf) [reply mlir/lib/Driver/CompilerDriver.cpp:390]: The dangling pipeline is basically an empty pipeline (the one with no passes in it) in the end of the pipeline list. We handle them here to maintain the invariant "on success, each pipeline must have its dump in the keep-intermediate mode". Without this code, empty trailing pipelines wouldn't be mentioned in dumps.

File: mlir/lib/Driver/CompilerDriver.cpp [hunk 3]
```diff
@@ -274,79 +286,108 @@ LogicalResult runEnzymePasses(const CompilerOptions &options,
...
-                s << moduleOp;
-            }
-            const std::string &outFile =
-                fs::path(options.moduleName.str()).replace_extension(".mlir");
-            if (failed(dumpToFile(options, outFile, tmp))) {
-                return failure();
-            }
+            llvm::raw_string_ostream s{tmp};
+            s << moduleOp;
         }
```
[the remark this answers is not in the document]
Comment (grwlf) [reply mlir/lib/Driver/CompilerDriver.cpp:339]: Are you also suggesting to remove the dumpFile call? I think we should limit the lifetime of `s` because we need it to flush the stream before we pass it to the dump function.

File: frontend/catalyst/compiler.py [hunk 4]
```diff
@@ -339,7 +339,13 @@ def run_from_ir(
...
                ret_type_name (str) Inferred main function result type name
         """
-        pipelines = pipelines if pipelines is not None else DEFAULT_PIPELINES
+        pipelines = (
+            pipelines
+            if pipelines is not None
+            else (
+                self.options.pipelines if self.options.pipelines is not None else DEFAULT_PIPELINES
+            )
+        )
```
[the remark this answers is not in the document]
Comment (grwlf) [reply frontend/catalyst/compiler.py:348]: Users want to specify the pipelines via the parameters, but we also use an utility pipeline consisting of a single canonicalize call. So I thought it is better to handle the logic closer to the point where we use it. Do you think we should resolve pipelines outside this function?

**Comment by grwlf** [comment]
Thanks for the review, Erick! Let me answer some questions

> Cool! I am not a fan of creating a pass that just fails in order to test the reporting framework. I think it is enough to have something similar to the "compile from IR" error where that it is guaranteed to fail at compile time. As to what is something that is guaranteed to fail at compile time that cannot be detected by a simple parse? I am not too sure. So perhaps, this test is indeed needed.

Yes, I am also not very sure, but I think it would be better to have something like a test pass where we can encode errors we want to test. It is always failing now, but I think we can add other modes via options. I expect we might need to test warnings as well, for example.

Also, as it turns out, it was not so easy for me to emit the right error using the "compile from IR" mode. Most of the obvious errors are in fact mlir parsing errors, while here we need logical ones (but not asserts!). So I decided just to add a tool for it. I am open to suggestions though.    

> 
> What are dangling pipelines?

Answered in a separate suggestion below

> It also looks like the print handler is called after each single pass inside the pipeline as opposed to just call the print handler at the end of the pipeline
> 
> ```c++
>     // For each pipeline-terminating pass, print the IR into the corresponding dump file and
>     // into a diagnostic output buffer. Note that one pass can terminate multiple pipelines.
>     auto afterPassCallback = [&](Pass *pass, Operation *op) {
>         if (!options.keepIntermediate)
>             return;
>         auto res = pipelineTailMarkers.find(pass);
>         if (res != pipelineTailMarkers.end()) { // <-- only if it is the end it is printed
>         // but the fact that we are in the printer callback is a bit strange to me.
> ```

I think it should be fine because the pass-printing API we used before does the same under the hood. I hooked a lower-level API to get access to the additional `afterPassFailedCallback` handler which we also need in order to print the pipeline name.

> I also feel like the printing mechanism has gotten a bit convoluted. 
> I am thinking something that instead looks like:
> 
> ```c++
> for pipeline {
>   // prepare pass manager
>    if (failed(pm.run(moduleOp))) {
>         return failure();
>     }
>     print moduleOp
> }
> ```
> 
> This is already being done for the LLVM pipelines. What do you think? Is it possible? I also feel like this would enable us to use multithreading since it appears that multithreading is disabled when printing using the IR printer as opposed to calling print on the module itself.

The current join-everything-into-one-pipeline approach was here before me, so I didn't want to touch it initially. But the changes are already quite significant so I think it might be worth trying. But note, that splitting a single `pm.run` that we have now into many runs might have some consequences. Here is the [source](https://github.com/llvm/llvm-project/blob/4706251a3186c34da0ee8fd894f7e6b095da8fdc/mlir/lib/Pass/Pass.cpp#L804). Affected concepts are dependent dialects, pass initialization, analysis manager, crash recovery and multi-threading.

File: frontend/catalyst/compiler.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply frontend/catalyst/compiler.py:348]: It is hard to follow given all the variables and fields named pipelines here.

I normally just prefer to have:

```python
if param is None:
     param = some_default
```

Why not reuse the single canonicalize call in `self.options.pipelines = canonicalize`? Or better, create a temporary CompilerOptions just for that canonicalize call.

Ideally this call is just like a library and different callers should not have to have different ways of having the same functionality.

File: mlir/lib/Driver/CompilerDriver.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply mlir/lib/Driver/CompilerDriver.cpp:339]: I see. I made a mistake, I deleted `dumpToFile`. Sorry about that!

I am looking at the [documentation](https://llvm.org/doxygen/classllvm_1_1raw__string__ostream.html) for `llvm::raw_string_ostream` and it says the following:

> This is a simple adaptor class. This class does not encounter output errors. [raw_string_ostream](https://llvm.org/doxygen/classllvm_1_1raw__string__ostream.html) operates without a buffer, delegating all memory management to the std::string. Thus the std::string is always up-to-date, may be used directly and there is no need to call [flush()](https://llvm.org/doxygen/classllvm_1_1raw__ostream.html#a520bdf57dfe3e73abb53d482893f0a27).

Comment (erick-xanadu) [inline mlir/lib/Driver/CompilerDriver.cpp:339]:
```suggestion
        llvm::raw_string_ostream s{tmp};
        s << moduleOp;
```

[the remark this answers is not in the document]
Comment (erick-xanadu) [reply mlir/lib/Driver/CompilerDriver.cpp:339]: > Are you also suggesting to remove the dumpToFile call? 

Looking back, I don't think I am suggesting to remove `dumpToFile`. `dumpToFile` is still below. All I'm suggesting is to remove the braces that limit the scope of `s`

[the remark this answers is not in the document]
Comment (grwlf) [reply mlir/lib/Driver/CompilerDriver.cpp:339]: Regarding dumpToFile - Yes, my bad, sorry. That was my own change.

Thanks for checking the docs, the scope is indeed not needed if it doesn't use buffers.  I will remove it soon.

File: mlir/lib/Driver/CompilerDriver.cpp (hunk 2 above)
Comment (erick-xanadu) [reply mlir/lib/Driver/CompilerDriver.cpp:390]: Do we need to support this use case? I feel like it would be best to just have a filtering step in the frontend where we remove the empty pipelines. My reasoning is this:

1. pipelines start with MLIR and end with MLIR.
2. If you need to print the output of the end of a dangling pipeline, then just print the input, which clearly we already have (the `.mlir` field as the base start or the end of the last pipeline in the case you describe above).

Comment (erick-xanadu) [reply mlir/lib/Driver/CompilerDriver.cpp:390]: What I concretely mean is adding the invariant that empty pipelines are effectively no-op compiler transformations including the printing of the IR.

Comment (grwlf) [reply mlir/lib/Driver/CompilerDriver.cpp]: I think that current design would result in a simpler user API regarding the dumping and getting pipeline output. The PR version allows users to get the correct output for any pipeline. You are effectively suggesting to change the API and ask users to check if their pipelines are empty or not by themselves. I am not strictly against it, the former was my default preference.

File: mlir/lib/Driver/CompilerDriver.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (grwlf) [reply mlir/lib/Driver/CompilerDriver.cpp]: Done

File: mlir/lib/Driver/CompilerDriver.cpp (hunk 2 above)
Comment (grwlf) [reply mlir/lib/Driver/CompilerDriver.cpp]: I see it in this way: the current suggestion allows users to rely on the fact that whenever they enable `keep_intermediate` they would see dumps for all pipelines they configure no matter what. Same for `get_output`. Pipeline contents might become empty as a result of some user-side automation, so I think it is better to maintain 1-to-1 correspondence between pipelines and results.
Your suggestion involving frontend-side checks may result in some outputs missing (trailing empty pipelines). Not a big deal, and I agree that if we implement it, the source code will look simpler on our side, so I am not entirely sure. But I think I would prefer the API simplicity in this case.
I am ready to implement your suggestion if you are not convinced by now.

File: frontend/catalyst/compiler.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (grwlf) [reply frontend/catalyst/compiler.py]: > if param is None:
>     param = some_default

Got it. I think I like oneliners because  typecheckers I used in other projects tend to issue better error messages on type mismatches. This is because Python requires programmers to always specify `else` branch in the `.. if .. else ..` expressions. For Catalyst this argument does not work. 

I have fixed this one by adding `pipelines` into the `CompilerOptions`. Should look better now, please check!

[the remark this answers is not in the document]
Comment (erick-xanadu) [reply frontend/catalyst/compiler.py]: Thank you.

File: mlir/lib/Driver/CompilerDriver.cpp (hunk 2 above)
Comment (erick-xanadu) [reply mlir/lib/Driver/CompilerDriver.cpp]: I am against dangling pipelines. I would strongly prefer their removal :sweat_smile: Thanks!

Comment (grwlf) [reply mlir/lib/Driver/CompilerDriver.cpp]: Done

File: frontend/catalyst/compiler.py [hunk 5]
```diff
@@ -307,28 +326,26 @@ def run(infile, outfile=None, flags=None, fallback_compilers=None, options=None)
...
-    def __init__(self, options: Optional[CompileOptions] = None):
-        self.options = options if options is not None else CompileOptions
+    def __init__(self, workspace: Directory, options: Optional[CompileOptions] = None):
+        self.options = options if options is not None else CompileOptions()
         self.last_compiler_output = None
+        assert isinstance(
+            workspace, Directory
+        ), f"Compiler expects a Directory type, got {type(workspace)}."
+        assert workspace.is_dir(), f"Compiler expects an existing directory, got {workspace}."
+        self.workspace = workspace
```
Comment (erick-xanadu) [inline frontend/catalyst/compiler.py:336]: Can we remove the workspace from the Compiler? I would like to have the compiler be as stateless as possible. Here is a comment that says that we shouldn't move the workspace to the Compiler class. The reason is that I see the Compiler class as just a command line and the workspace is the `-o` flag with a specific path.

https://github.com/PennyLaneAI/catalyst/pull/303/files#diff-d57fb4415becbb6d125b231ba48c92d4df9bf73c43c32583889555883041e466R484-R497

Comment (grwlf) [reply frontend/catalyst/compiler.py]: Oh I see. I have reverted this part, thanks!
